### PR TITLE
Support CUDAGraph

### PR DESCRIPTION
**Pull request by hogura99**

### CODE REVIEW

File: disagmoe/frontend/engine.py [hunk 1]
```diff
@@ -215,6 +226,77 @@ def setup_engine(self,
...
+    def _create_cuda_graph_contexts(self):
+        assert self.model_config.enable_cuda_graph
+        self.graphs: List[List[torch.cuda.CUDAGraph]] = []
+        self.static_input = torch.zeros((MAX_BATCH_SIZE, self.model_config.hidden_size)).to("cuda", non_blocking=True)
+        self.static_output = torch.zeros((MAX_BATCH_SIZE, self.model_config.hidden_size)).to("cuda", non_blocking=True)
+        if self.is_attn:
+            self.static_expert_ids = torch.zeros((MAX_BATCH_SIZE, ), dtype=torch.long).to("cuda", non_blocking=True)
+            self.static_positions = torch.zeros(MAX_BATCH_SIZE, dtype=torch.long).to("cuda", non_blocking=True)
+            self.static_block_table = torch.zeros(
+                (MAX_BATCH_SIZE, self.cache_config.num_gpu_blocks),
```
Comment (shawlleyw) [inline disagmoe/frontend/engine.py:238]: block table shape should be (batch size, max length / block size)

File: disagmoe/frontend/engine.py [hunk 2]
```diff
@@ -257,16 +343,34 @@ def _pack_flash_attn_metadata(
...
+            # mocking=True when _warmup_attn
+            block_table_1d = torch.zeros(
+                (num_tokens + num_seqs * MAX_SEQ_LEN // self.cache_config.block_size, ), 
+                dtype=torch.int32).to("cuda")
+        
+        if self.model_config.enable_cuda_graph:
+            self.static_slot_mapping[0: num_seqs].copy_(
+                block_table_1d[-num_seqs : ].to(torch.int64))
+            _max_num_blocks = (block_table_1d.numel() - num_seqs) // num_tokens
+            self.static_block_table[0: num_tokens, 0: _max_num_blocks].copy_(
```
Comment (shawlleyw) [inline disagmoe/frontend/engine.py:366]: We can still use static block table as a 1D tensor, while passing it to executor with "viewing" as 2D tensor

File: disagmoe/frontend/engine.py [hunk 3]
```diff
@@ -621,7 +614,7 @@ def process_batch_expert(self,
         batch_sizes = torch.tensor(
             [batch_sizes[i] for i in self.inner_exp_rank],
             dtype=torch.int64,
-            device="cpu",   # NOTE(hogura|20241014): grouped_gemm requires batch_sizes to be on cpu
+            device="cuda",   # NOTE(hogura|20241014): grouped_gemm requires batch_sizes to be on cpu
```
Comment (hogura99) [inline disagmoe/frontend/engine.py]: Maybe leave an option to decide whether CUBLAS or CUTLASS?

**Comment by shawlleyw** [comment]
Now we have not supported CUDA graph for experts.

**Pull request merged**